README.md
[詳細はコチラ](http://atocrafter.github.io/ThreeMeals/)


ChangeLog.txt
# 更新履歴


## バージョン別

### 1.0.0

リリース。Minecraft 1.6.4 用。


src/main/java/ato/threemeals/ThreeMeals.java
package ato.threemeals;

import cpw.mods.fml.common.Mod;
import cpw.mods.fml.common.event.FMLInitializationEvent;
import cpw.mods.fml.common.event.FMLPreInitializationEvent;
import cpw.mods.fml.common.registry.GameRegistry;
import net.minecraft.block.Block;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraftforge.common.Configuration;
import net.minecraftforge.common.MinecraftForge;

@Mod(modid = "ato.threemeals")
public class ThreeMeals {
    private int blockBathroomScaleId;

    @Mod.EventHandler
    public void preInit(FMLPreInitializationEvent event) {
        loadConfig(event);
    }

    @Mod.EventHandler
    public void init(FMLInitializationEvent event) {
        MinecraftForge.EVENT_BUS.register(new PlayerEventHandler());

        String unlocalizedName = "bathroomscale";
        Block bathroomScale = new BlockBathroomScale(blockBathroomScaleId)
                .setUnlocalizedName(unlocalizedName)
                .setCreativeTab(CreativeTabs.tabRedstone);
        GameRegistry.registerBlock(bathroomScale, unlocalizedName);
        GameRegistry.addRecipe(new ItemStack(bathroomScale), new Object[]{
                "EE",
                'E', Item.emerald
        });
    }

    /**
     * コンフィグファイルを読み込む
     */
    private void loadConfig(FMLPreInitializationEvent event) {
        Configuration config = new Configuration(event.getSuggestedConfigurationFile());
        config.load();

        HardcoreFoodStats.FOODEXHAUSTION_SPEED = config.get(config.CATEGORY_GENERAL, "FoodExhaustionSpeed", 3).getInt(3);
        HardcoreFoodStats.HARAHERING_SPEED = config.get(config.CATEGORY_GENERAL, "HaraheringSpeed", 1).getInt(1);
        HardcoreFoodStats.ALWAYS_HARDMODE = config.get(config.CATEGORY_GENERAL, "AlwaysHardmode", true).getBoolean(true);

        blockBathroomScaleId = config.getBlock("BathroomScale", 3639).getInt(3639);

        config.save();
    }
}


src/main/java/ato/threemeals/BlockBathroomScale.java
package ato.threemeals;

import net.minecraft.block.BlockBasePressurePlate;
import net.minecraft.block.material.Material;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.world.World;

import java.util.List;

public class BlockBathroomScale extends BlockBasePressurePlate {
    protected BlockBathroomScale(int id) {
        super(id, "emerald_block", Material.iron);
        setHardness(0.5F);
        setStepSound(soundMetalFootstep);
    }

    @Override
    protected int getPlateState(World world, int i, int j, int k) {
        float weight = 0;

        List<Entity> list = world.getEntitiesWithinAABB(EntityPlayer.class, this.getSensitiveAABB(i, j, k));
        if (list != null && !list.isEmpty()) {
            for (Entity entity : list) {
                if (entity instanceof EntityPlayer) {
                    weight += ((HardcoreFoodStats)((EntityPlayer) entity).getFoodStats()).getFatness();
                }
            }
        }

        return (int) Math.min(weight / 10, 15);
    }

    @Override
    protected int getPowerSupply(int i) {
        return i;
    }

    @Override
    protected int getMetaFromWeight(int i) {
        return i;
    }
}


src/main/java/ato/threemeals/PlayerEventHandler.java
package ato.threemeals;

import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.potion.PotionEffect;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.event.ForgeSubscribe;
import net.minecraftforge.event.entity.living.LivingEvent;

public class PlayerEventHandler {
    private float fatness;
    private float starveness;

    /**
     * 肥満による悪影響
     */
    @ForgeSubscribe
    public void fatDebuff(LivingEvent.LivingUpdateEvent event) {
        if (event.entityLiving instanceof EntityPlayerMP) {
            EntityPlayerMP player = (EntityPlayerMP) event.entityLiving;
            if (player.getFoodStats() instanceof HardcoreFoodStats && !player.worldObj.isRemote) {
                HardcoreFoodStats stats = (HardcoreFoodStats) player.getFoodStats();
                fatness = stats.getFatness();
                starveness = stats.getStarveness();

                if (200 < stats.getFatness()) {
                    int amplifier = (int) ((stats.getFatness() - 200) / 50);
                    player.addPotionEffect(new PotionEffect(2, 10 * 20, amplifier));
                }
                if (400 < stats.getFatness()) {
                    player.addPotionEffect(new PotionEffect(8, 10 * 20, 128));
                }
            }
        }
    }

    /**
     * 空腹度に関する情報を表示
     */
    @ForgeSubscribe
    public void renderThreeMealsInfo(RenderGameOverlayEvent.Text event) {
        if (Minecraft.getMinecraft().gameSettings.showDebugInfo) {
            event.left.add("Fatness: " + fatness);
            event.left.add("Starveness: " + starveness);
        }
    }
}


src/main/java/ato/threemeals/asm/FoodStatsTransformer.java
package ato.threemeals.asm;

import ato.threemeals.HardcoreFoodStats;
import net.minecraft.launchwrapper.IClassTransformer;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.*;

import java.util.ListIterator;

public class FoodStatsTransformer implements IClassTransformer {
    @Override
    public byte[] transform(String name, String transformedName, byte[] basicClass) {
        if ("net.minecraft.entity.player.EntityPlayer".equals(transformedName)) {
            return patchFoodStats(basicClass);
        } else {
            return basicClass;
        }
    }

    private byte[] patchFoodStats(byte[] bytes) {
        ClassNode cnode = new ClassNode();
        ClassReader creader = new ClassReader(bytes);
        creader.accept(cnode, 0);

        for (MethodNode m : cnode.methods) {
            if ("<init>".equals(m.name)) {
                ListIterator<AbstractInsnNode> ite = m.instructions.iterator();
                while (ite.hasNext()) {
                    AbstractInsnNode insn = ite.next();
                    if (insn.getOpcode() == Opcodes.NEW && insn instanceof TypeInsnNode) {
                        TypeInsnNode newInsn = (TypeInsnNode) insn;
                        if ("net/minecraft/util/FoodStats".equals(newInsn.desc) ||
                                "ux".equals(newInsn.desc)) {
                            newInsn.desc = Type.getInternalName(HardcoreFoodStats.class);
                            ((MethodInsnNode) insn.getNext().getNext()).owner = Type.getInternalName(HardcoreFoodStats.class);
                        }
                    }
                }
            }
        }

        ClassWriter cwriter = new EntityPlayerClassWriter(ClassWriter.COMPUTE_FRAMES | ClassWriter.COMPUTE_MAXS);
        cnode.accept(cwriter);
        return cwriter.toByteArray();
    }
}


src/main/java/ato/threemeals/HardcoreFoodStats.java
package ato.threemeals;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.FoodStats;
import net.minecraft.world.World;

public class HardcoreFoodStats extends FoodStats {
    /**
     * 腹が減るスピード、バニラからの倍率
     * Config から変更可能
     */
    public static float FOODEXHAUSTION_SPEED;
    /**
     * 時間経過で腹が減るスピード
     * Config から変更可能
     */
    public static float HARAHERING_SPEED;
    /**
     * 常に難易度がハードモードで空腹度管理にするか
     * Config から変更可能
     */
    public static boolean ALWAYS_HARDMODE;
    /**
     * 肥満度
     * 多いほど、死ににくく燃費が悪くなる
     * [0, 100]: 健康
     * (100, 200]: ぽっちゃり, 燃費が悪い
     * (200, 400]: 肥満, 移動速度低下
     * (400, ): 超肥満, 常時ジャンプできない
     */
    protected float fatness = 0;
    /**
     * 飢餓度 (%)
     * 多いほど、脂肪がつきやすくなる
     */
    protected float starveness = 100;

    @Override
    public void addStats(int foodLevel, float foodSaturationModifier) {
        // 食事による肥満度の増加
        fatness += foodLevel * (starveness / 100) / 10;                         // 栄養の約 10% は常に脂肪生成に回される
        fatness += Math.max(0, getFoodLevel() + foodLevel - 20) * (starveness / 100);   // 満腹度の超過分が脂肪になる
        fatness += Math.max(0, getSaturationLevel() + foodLevel * foodSaturationModifier * 2
                - Math.min(getFoodLevel() + foodLevel, 20)) * (starveness / 100);       // 隠し満腹度の超過分が脂肪になる

        super.addStats(foodLevel, foodSaturationModifier);
    }

    @Override
    public void onUpdate(EntityPlayer entityPlayer) {
        if (ALWAYS_HARDMODE) {
            World world = entityPlayer.worldObj;
            int difficultyBackup = world.difficultySetting;
            world.difficultySetting = 3;
            super.onUpdate(entityPlayer);
            world.difficultySetting = difficultyBackup;
        } else {
            super.onUpdate(entityPlayer);
        }
        harahering();
        // 飢餓度の時間経過による減少
        starveness *= 0.99999;  // 半減期が約 1 時間 (= 3 日)
    }

    @Override
    public void readNBT(NBTTagCompound nbttc) {
        if (nbttc.hasKey("fatness")) {
            fatness = nbttc.getFloat("fatness");
        }
        if (nbttc.hasKey("starveness")) {
            starveness = nbttc.getFloat("starveness");
        }
        super.readNBT(nbttc);
    }

    @Override
    public void writeNBT(NBTTagCompound nbttc) {
        nbttc.setFloat("fatness", fatness);
        nbttc.setFloat("starveness", starveness);
        super.writeNBT(nbttc);
    }

    /**
     * 時間経過腹減りモード
     * 何もしなくても時間経過で腹が減る
     */
    private void harahering() {
        float haraheri = HARAHERING_SPEED / 1000;
        if (getFoodLevel() < 5) {
            // 脂肪による飢餓の回避
            double hungry = Math.min((5 - getFoodLevel()) * 0.25, 1);
            fatness = (float) Math.max(0, fatness - haraheri * hungry);
            if (fatness > 0) haraheri *= 1 - hungry;
            starveness = (float) Math.min(starveness + Math.pow(hungry, 2) / 1000, 100);
        }
        addExhaustion(haraheri);
    }

    @Override
    public void addExhaustion(float exhaustion) {
        // 運動による脂肪の燃焼
        fatness = (float) Math.max(0, fatness - Math.pow(exhaustion, 2) * 2);
        // ハードコア腹減りモード
        // FOODEXHAUSTION_SPEED 倍の速度で腹が減る
        exhaustion *= FOODEXHAUSTION_SPEED;
        // 肥満度による腹減り速度増加
        exhaustion *= Math.max(fatness / 100, 1);

        super.addExhaustion(exhaustion);
    }

    public float getStarveness() {
        return starveness;
    }

    public float getFatness() {
        return fatness;
    }
}
